include/libtorrent/aux_/win_crypto_provider.hpp

```diff
@@ -0,0 +1,170 @@
+			from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WIN_CRYPTO_PROVIDER_HPP
+#define TORRENT_WIN_CRYPTO_PROVIDER_HPP
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+
+#include <windows.h>
+#include <wincrypt.h>
+
+namespace libtorrent { namespace aux
+{
+
+	static HCRYPTPROV crypt_acquire_provider(DWORD provider_type)
+	{
+		HCRYPTPROV ret;
+		if (CryptAcquireContext(&ret, nullptr, nullptr, provider_type
+			, CRYPT_VERIFYCONTEXT) == false)
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+		return ret;
+	}
+
+	static void crypt_gen_random(span<char> buffer)
+	{
+		static HCRYPTPROV provider = crypt_acquire_provider(PROV_RSA_FULL);
+		if (!CryptGenRandom(provider, int(buffer.size())
+			, reinterpret_cast<BYTE*>(buffer.data())))
+		{
+#ifndef BOOST_NO_EXCEPTIONS
+			throw system_error(error_code(GetLastError(), system_category()));
+#else
+			std::terminate();
+#endif
+		}
+	}
+
+	template <ALG_ID AlgId, DWORD ProviderType>
+	struct crypt_hash
+	{
+		crypt_hash() { create(); }
+		crypt_hash(crypt_hash const& h) { duplicate(h); }
+		~crypt_hash() { destroy(); }
+
+		crypt_hash& crypt_hash::operator=(crypt_hash const& h)
+		{
+			if (this == &h) return *this;
+			destroy();
+			duplicate(h);
+			return *this;
+		}
+
+		void update(span<char const> data)
+		{
+			TORRENT_ASSERT(m_initialized);
+
+			if (CryptHashData(m_hash, reinterpret_cast<BYTE const*>(data.data()), int(data.size()), 0) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+		}
+
+		void get_hash(char *digest, size_t digest_size)
+		{
+			TORRENT_ASSERT(m_initialized);
+
+			DWORD size = DWORD(digest_size);
+			if (CryptGetHashParam(m_hash, HP_HASHVAL
+				, reinterpret_cast<BYTE*>(digest), &size, 0) == false)
+			{
+#ifndef BOOST_NO_EXCEPTIONS
+				throw system_error(error_code(GetLastError(), system_category()));
+#else
+				std::terminate();
+#endif
+			}
+			TORRENT_ASSERT(size == DWORD(digest_size));
+		}
+	private:
+		void create()
```

Comment:
this function is only called from the constructor. is there a reason to not just make it be the constructor?